Fix update field_attributes for Account Model

hongthang152 opened the pull request:
#9611 
Remove checking for Hash object, change to checking for Array object.

hongthang152 committed:
Fix update field_attributes for Account Model

Remove checking for Hash object, change to checking for Array object.

hongthang152 committed:
Change kind_of? to is_a? for checking Array

hongthang152 committed:
Update account.rb

Gargron commented on app/models/account.rb:
```diff
@@ -230,8 +230,8 @@ def fields_attributes=(attributes)
 
-    if attributes.is_a?(Hash)
-      attributes.each_value do |attr|
+    if attributes.is_a?(Array)
```
This change isn't right either, because we get a hash from a standard form submission. So it can be a hash or an array.

hongthang152 replied:
Change to 
`Array.wrap(attributes).each do |attr|`

We wrap the array around a single element to allow iteration.

hongthang152 replied:
@Gargron You can check the latest commit

hongthang152 committed:
Remove checking for Array. 

Wrap Array around single hash object and iterate over to update custom profile fields.

hongthang152 committed:
Clean up checking Hash object 

Also apply next to skip iteration

nightpool commented on app/models/account.rb:
```diff
@@ -230,18 +230,16 @@ def fields_attributes=(attributes)
-      attributes.each_value do |attr|
-        next if attr[:name].blank?
+    Array.wrap(attributes).each do |attr|
+      next if !attr.is_a?(Hash) || attr[:name].blank?
```
we should throw a validation error here instead of passing silently, to resolve #9611

hongthang152 replied:
It raises error now when it fails the validation.

hongthang152 replied:
```
     if !attr.is_a?(Hash) || attr[:name].blank?
        raise ArgumentError, 'One of custom field attributes has invalid format'
      end
```

hongthang152 committed:
Raise ArgumentError if validation fails

hongthang152 committed:
Remove trailing white space

Gargron commented on app/models/account.rb:
```diff
@@ -230,18 +230,17 @@ def fields_attributes=(attributes)
-        end
+    Array.wrap(attributes).each do |attr|
+      if !attr.is_a?(Hash) || attr[:name].blank?
+        raise ArgumentError, 'One of custom field attributes has invalid format'
```
I'm sorry, I disagree with @nightpool on this... Raising ArgumentError will lead to error 500 in the API, so it won't be very useful. It's possible to raise ActiveRecord::RecordInvalid, but I am not 100% sure if this is the right place to do it (this is in a setter, not in the validation cycle...)

Until there is a better validation solution, silently ignoring seems better. Sorry for long delay.
